src/validation.cpp

```diff
@@ -862,7 +863,62 @@ static bool WriteBlockToDisk(const CBlock &block, FlatFilePos &pos,
 
 Amount GetBlockSubsidy(uint32_t nBits,
                        const Consensus::Params &consensusParams) {
-    return SUBSIDY;
+    static constexpr int64_t SUBSIDY_INT = SUBSIDY / SATOSHI;
+    if (!consensusParams.enableDifficultyBasedSubsidy) {
+        return SUBSIDY;
+    }
+    // We define:
+    // R: Reward
+    // D: Difficulty
+    // T: Target
+    // Tm: Max Target
+    // a: Base subsidy
+
+    // In Lotus, the reward R = a*log(D)
+    // R = a*log(Tm / T) = a*log(Tm) - a*log(T)
+
+    // We get the target by decomposing nBits:
+    // T = mantissa * 2^(exp - 3)
+
+    // The log of that cancels out the 2^x
+    // log(T) = log(mantissa) + exp - 3
+
+    // We can approximate log(x) using Log2FixedPoint.
+    // log(x) = Log2FixedPoint(2^b * x) / 2^b, where b is the precision.
+    // We use 16 bits for the precision.
+    static constexpr int64_t PRECISION_BITS = 16;
+
+    // log of max target, rounded up. It actually is 223.99997798605273, we
+    // correct for this by calculating with mantissa + 1.
+    static constexpr int64_t LOG_MAX_TARGET = 224;
+    const int32_t exp = nBits >> 24;
+    const uint32_t mantissa = (nBits & 0x007fffff);
+
+    // Compute a*log(nMantissa)
+    // We re-interpret mantissa as fixed point number with 16 bits in the
```

Comment:
We're not reinterpreting the mantissa as a fixed point, rather Log2FixedPoint gives back a number in fixed point.

Reply:
No, we are. Log2FixedPoint expects a fixed point number, but we provide a normal integer. Log2FixedPoint will treat it as fixed point, so we correct for that later.